File: PrimaMateria/PrimaMateria/Classes/XTRElementListViewController.m
//
//  XTRElementListViewController.m
//  PrimaMateria
//
//  Created by Jerry Porter on 11/21/11.
//  Copyright (c) 2016 xTrensa. All rights reserved.
//

@interface XTRElementListViewController () {
    UITableView *tableView;
}
- (void)setupTableView;
- (void)showElementPanelForElementAtIndex: (int) anIndex;
@end

@implementation XTRElementListViewController

#pragma mark Private Methods

- (void)showElementPanelForElementAtIndex: (int) anIndex {
    [XTRPropertiesStore storeViewTitle: self.title];
    [XTRPropertiesStore storeAtomicNumber:@(anIndex)];
    [self performSegueWithIdentifier: SHOW_INSPECTOR_FROM_ELEMENT_LIST sender: self];
}

- (void)setupTableView {
    tableView = [[UITableView alloc] initWithFrame:self.swapView.frame style: UITableViewStylePlain];
    tableView.alwaysBounceVertical = NO;
    tableView.alwaysBounceHorizontal = NO;
    tableView.dataSource = self;
    tableView.delegate = self;
    tableView.separatorStyle = UITableViewCellSeparatorStyleNone;
    tableView.backgroundColor = UIColor.blackColor;
    [self.view addSubview: tableView];
}

#pragma mark - Action Methods

- (IBAction) sortTableView: (id) sender {
    BOOL toggleState = [sender toggleState];
    
    if (tableView != nil) {
        [tableView removeFromSuperview];
        tableView = nil;
    }
    
    [[XTRDataSource sharedInstance] sortByColumnPosition:[sender tag] andOrdering: toggleState];
    [self setupTableView];
    [tableView reloadData];
}

#pragma mark - UITableView DataSource Methods

- (UITableViewCell *) tableView: (UITableView *) aTableView cellForRowAtIndexPath: (NSIndexPath *) indexPath {
    NSInteger row = indexPath.row;
    XTRElement *element = [[XTRDataSource sharedInstance] sortedElementAtIndex: row];
    XTRElementTableViewCell *cell = [aTableView dequeueReusableCellWithIdentifier: @"XTRElementTableViewCell"];
    
    if (cell == nil) {
        cell = [[XTRElementTableViewCell alloc] initWithStyle: UITableViewCellStyleDefault reuseIdentifier: @"XTRElementTableViewCell" andElement: element];
    }
    
    [cell modifyCellProperties];
    
    return cell;
}

- (NSInteger) numberOfSectionsInTableView: (UITableView *) aTableView {
    return 1;
}

- (NSInteger) tableView: (UITableView *) aTableView numberOfRowsInSection: (NSInteger) section {
    return [XTRDataSource sharedInstance].sortedElementList.count;
}

#pragma mark - UITableView Delegate Methods

- (void)tableView: (UITableView *)aTableView didSelectRowAtIndexPath: (NSIndexPath *) indexPath {
    XTRElement *element = [[XTRDataSource sharedInstance] sortedElementAtIndex:(NSUInteger) indexPath.row];
    [aTableView deselectRowAtIndexPath: indexPath animated: YES];
    [self showElementPanelForElementAtIndex: (element.atomicNumber.intValue - 1)];
}

- (CGFloat)tableView:(UITableView *)aTableView heightForRowAtIndexPath:(NSIndexPath *)indexPath {
    return 42.0;
}

#pragma mark - View Management Methods

- (void)viewDidLoad {
    [super viewDidLoad];
    
    [self.atomicNumberButton toggleState];
    [self.swapView removeFromSuperview];
    [self setupTableView];
}

- (BOOL) shouldAutorotateToInterfaceOrientation: (UIInterfaceOrientation) interfaceOrientation {
    return UIInterfaceOrientationIsLandscape(interfaceOrientation);
}


#pragma mark - Memory Management Methods

- (void)dealloc {
    self.atomicNumberButton = nil;
    self.swapView = nil;
}

@end

File: PrimaMateria/PrimaMateria/Classes/XTRElementListViewController.swift
//
//  XTRElementListViewController.swift
//  PrimaMateria
//
//  Created by Jerry Porter on 4/15/16.
//  Copyright ©2018 xTrensa. All rights reserved.
//

class XTRElementListViewController : UIViewController, UITableViewDelegate, UITableViewDataSource {
    
    static let tableViewCellIdentifier = "XTRElementTableViewCell"
    
    @IBOutlet var atomicNumberButton : XTRTableHeaderButton!
    @IBOutlet var swapView: UIView!
    
    private var tableView : UITableView?
    private var _indexPath : IndexPath?
    
    // MARK: - Initialization Methods
    
    required init?(coder aDecoder: NSCoder) {
        super.init(coder: aDecoder)!
    }
    
    // MARK: - Internal Methods
    
    private func showElementPanelForElementAtIndex(_ anIndex: Int)  {
        XTRPropertiesStore.viewTitle = title!
        XTRPropertiesStore.atomicNumber = anIndex
        performSegue(withIdentifier: SegueName.showInspectorFromElementList, sender: self)
    }
    
    private func setupTableView() {
        tableView = UITableView(frame: swapView.frame, style: .plain)
        guard let localTableView = tableView else { return }
        localTableView.alwaysBounceVertical = false
        localTableView.alwaysBounceHorizontal = false
        localTableView.delegate = self
        localTableView.dataSource = self
        localTableView.separatorStyle = .none
        localTableView.backgroundColor = UIColor.black
        localTableView.allowsSelection = true
        localTableView.allowsMultipleSelection = false
        localTableView.register(XTRElementTableViewCell.classForCoder(), forCellReuseIdentifier: XTRElementListViewController.tableViewCellIdentifier)
        view.addSubview(tableView!)
    }
    
    // MARK: - Action Methods
    
    @IBAction func sortTableView(_ sender: XTRTableHeaderButton) {
        let toggleState = sender.toggleState()
        
        if tableView != nil {
            tableView!.delegate = nil
            tableView!.dataSource = nil
            tableView!.removeFromSuperview()
            tableView = nil
        }
        
        XTRDataSource.sharedInstance().sortByColumnPosition(sender.tag, andOrdering: toggleState)
        setupTableView()
        tableView!.reloadData()
    }
    
    // MARK: - View Management Methods
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        _ = atomicNumberButton.toggleState()
        swapView.removeFromSuperview()
        setupTableView()
        navigationController?.navigationBar.prefersLargeTitles = true
    }
    
    override func viewDidDisappear(_ animated: Bool) {
        if _indexPath != nil  {
            tableView!.deselectRow(at: _indexPath!, animated: true)
        }
        super.viewDidDisappear(animated)
    }
    
    override var shouldAutorotate : Bool {
        return false
    }
    
    override var supportedInterfaceOrientations : UIInterfaceOrientationMask {
        return .landscape
    }
    
    // MARK: - Memory Management Methods
    
    deinit {
        atomicNumberButton = nil
        swapView = nil
    }
    
}

extension XTRElementListViewController { // TableView DataSource - Delegate
    // MARK: - UITableView DataSource Methods
    
    func tableView(_ aTableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let cell : XTRElementTableViewCell? = aTableView.dequeueReusableCell(withIdentifier: XTRElementListViewController.tableViewCellIdentifier) as? XTRElementTableViewCell
        cell!.setup(XTRDataSource.sharedInstance().sortedElementAtIndex(indexPath.row))
        
        return cell!
    }
    
    func numberOfSections(in aTableView: UITableView) -> Int {
        return 1
    }
    
    func tableView(_ aTableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return XTRDataSource.sharedInstance().sortedElementList!.count
    }
    
    // MARK: - UITableView Delegate Methods
    
    func tableView(_ aTableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        let element = XTRDataSource.sharedInstance().sortedElementAtIndex(indexPath.row)
        let atomicNumber = element.atomicNumber
        
        _indexPath = indexPath
        aTableView.deselectRow(at: indexPath, animated: true)
        showElementPanelForElementAtIndex(atomicNumber - 1)
    }
    
    func tableView(_ aTableView: UITableView, heightForRowAt indexPath: IndexPath) -> CGFloat {
        return 44
    }

}
